Update to spack-stack 1.6.0

DavidHuber-NOAA opened the pull request:
# Description
This updates the module files for the global workflow to use spack-stack 1.6.0 and upgrades the submodules ~~(except GDAS)~~ to the same versions.  This also addresses a couple minor issues discovered during testing: a bug in the `awips_g2` job where `fcsthrs` was undeclared, `build_upp.sh` can now accept the `-j #` flag to specify the number of build jobs, an optimization was made to `build_all.sh` to build the UFS with more build jobs which speeds up the overall build process, and inconsistencies between `build` and `run` version files were corrected.

This update allows two major upgrades to take place simultaneously:
1) The reenabling of support for MET and METplus, so EMC_verif-global support has been reinstated and updated on all platforms, including WCOSS2.
2) The upgrade of CRTM to version 2.4.0.1, which now aligns with the production version to take advantage of new data sources.

With spack-stack v1.6.0, the name of the gsi-addon environment is different across platforms, so the version files have been restructured to construct the path to the `Core` module files.

The libraries that were upgraded in this upgrade that affect the global workflow are
crtm 2.4.0 -> 2.4.0.1
netcdf-fortran 4.6.0 -> 4.6.1
cdo 2.0.5 -> 2.2.0 (except on Orion)
prod_util 1.2.2 -> 2.1.1
py-yaml 5.4.1 -> 6.0.0
xarray 2022.3.0 -> 2023.7.0

The new libraries used in the global workflow to support EMC_verif-global are
pandas v1.5.3
python_dateutil v2.8.2
met v9.1.3
metplus v3.1.1

Resolves #2195 #1734 #1575
Depends on
- ~~https://github.com/ufs-community/ufs-weather-model/pull/2093~~ The UFS upgrade will be tackled at a later time.
- https://github.com/NOAA-EMC/UPP/pull/856
- https://github.com/NOAA-EMC/GSI/pull/684
- https://github.com/NOAA-EMC/gfs-utils/pull/43
- https://github.com/NOAA-EMC/GSI-utils/pull/27
- https://github.com/NOAA-EMC/GSI-Monitor/pull/111
- https://github.com/ufs-community/UFS_UTILS/pull/892
- https://github.com/JCSDA/spack-stack/issues/952
- https://github.com/NOAA-EMC/EMC_verif-global/issues/117
- https://github.com/NOAA-EMC/GDASApp/pull/873
- https://github.com/JCSDA/spack-stack/issues/963
- https://github.com/NOAA-EMC/EMC_verif-global/pull/121

# Type of change
<!-- Delete all except one -->
- Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?

# Checklist
- [ ] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

DavidHuber-NOAA committed:
Update modulefiles, submodules to spack-stack 1.6.0. #2195

DavidHuber-NOAA committed:
Corrected gsi-addon-dev spelling.

DavidHuber-NOAA committed:
Corrected the SS env. name in the version files. #2195

DavidHuber-NOAA committed:
Merge remote-tracking branch 'origin/develop' into ss160

DavidHuber-NOAA commented:
@malloryprow FYI I am starting to test this in cycled mode.  Do you have an EMC_verif-global branch you would like me to include?

malloryprow commented:
> @malloryprow FYI I am starting to test this in cycled mode. Do you have an EMC_verif-global branch you would like me to include?

Awesome! Yes, it is https://github.com/NOAA-EMC/EMC_verif-global/tree/gw_updates, commit https://github.com/NOAA-EMC/EMC_verif-global/commit/1e7baf7982c81067c36768444d5f73e294862911

DavidHuber-NOAA committed:
Reenable verif-global. #2195

DavidHuber-NOAA committed:
Better optimize build jobs (more ufs jobs).

DavidHuber-NOAA commented:
@CoryMartin-NOAA @RussTreadon-NOAA Would you like to add an updated GDASApp hash that includes spack-stack 1.6.0 support to this PR?  I gave it a quick try myself, but there are a couple of libraries that are not available in 1.6.0, so I'm not sure if these need to be requested or if there are appropriate substitutes:

```
libunistring/1.1
libidn2/2.3.4
```

My attempt is here: `/scratch1/NCEPDEV/global/David.Huber/GW/gw_160/sorc/gdas.cd/modulefiles/GDAS/hera.lua`.

RussTreadon-NOAA commented:
It would be nice to get GDASApp to spack-stack/1.6.0 but I'm not sure if JEDI repos are ready for this.  What do you think @CoryMartin-NOAA .   @DavidHuber-NOAA , I will look at your `hera.lua`.

DavidHuber-NOAA committed:
Fix gsi-addon paths for hercules, orion, and S4.

DavidHuber-NOAA committed:
More gsi-addon path fixes.

CoryMartin-NOAA commented:
> It would be nice to get GDASApp to spack-stack/1.6.0 but I'm not sure if JEDI repos are ready for this. What do you think @CoryMartin-NOAA . @DavidHuber-NOAA , I will look at your `hera.lua`.

I'm not sure. I have no reason to suspect they aren't ready but I also haven't seen JCSDA migrate to using 1.6 for Skylab yet.

RussTreadon-NOAA commented:
While I was able to build GDASApp `develop` using the spack-stack/1.6.0 `hera.lua`, several GDASApp ctests failed

```
83% tests passed, 5 tests failed out of 30

Label Time Summary:
gdas-utils    =   6.47 sec*proc (9 tests)
script        =   6.47 sec*proc (9 tests)

Total Test time (real) = 113.09 sec

The following tests FAILED:
        1682 - test_gdasapp_jedi_increment_to_fv3 (Failed)
        1692 - test_gdasapp_soca_nsst_increment_to_mom6 (Failed)
        1693 - test_gdasapp_land_create_ens (Failed)
        1694 - test_gdasapp_land_imsproc (Failed)
        1696 - test_gdasapp_land_letkfoi_snowda (Failed)
```

Rebuild `develop` using `spack-stack-1.5.1`.  Ran ctests.  All tests passed.
```

100% tests passed, 0 tests failed out of 30

Label Time Summary:
gdas-utils    =  10.91 sec*proc (9 tests)
script        =  10.91 sec*proc (9 tests)

Total Test time (real) = 225.75 sec
```

RussTreadon-NOAA commented:
The failed `test_gdasapp` ctests from the spack-stack/1.6.0 build were rerun with `-VV`.  Each test has a common error message

```
1692:   File "/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/unified-env/install/intel/2021.5.0/py-numpy-1.22.3-fywzyg3/lib/python3.10/site-package\
s/numpy/core/overrides.py", line 6, in <module>
1692:     from numpy.core._multiarray_umath import (
1692: ModuleNotFoundError: No module named 'numpy.core._multiarray_umath'
```
followed by
```
1692: IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!
1692:
1692: Importing the numpy C-extensions failed. This error can happen for
1692: many reasons, often due to issues with your setup or how NumPy was
1692: installed.
1692:
1692: We have compiled some common reasons and troubleshooting tips at:
1692:
1692:     https://numpy.org/devdocs/user/troubleshooting-importerror.html
1692:
1692: Please note and check the following:
1692:
1692:   * The Python version is: Python3.7 from "/scratch1/NCEPDEV/da/python/opt/core/miniconda3/4.6.14/envs/gdasapp/bin/python3"
1692:   * The NumPy version is: "1.22.3"
1692:
1692: and make sure that they are the versions you expect.
1692: Please carefully study the documentation linked above for further help.
1692:
1692: Original error was: No module named 'numpy.core._multiarray_umath'
```

Not sure where to go with this but I'll keep poking around.

CoryMartin-NOAA commented:
Looks like the python version used by the ctests is not the correct one. @guillaumevernieres had a similar issue with his AI tinkering on Hera. We may need to do the following in the cmake command: ``-DPYTHON_EXECUTABLE=`which python```

RussTreadon-NOAA commented:
I made the following change to `GDASApp` `build.sh`
```
     module load GDAS/$BUILD_TARGET
-    CMAKE_OPTS+=" -DMPIEXEC_EXECUTABLE=$MPIEXEC_EXEC -DMPIEXEC_NUMPROC_FLAG=$MPIEXEC_NPROC -DBUILD_GSIBEC=ON"
+    CMAKE_OPTS+=" -DMPIEXEC_EXECUTABLE=$MPIEXEC_EXEC -DMPIEXEC_NUMPROC_FLAG=$MPIEXEC_NPROC -DBUILD_GSIBEC=ON -DPYTHON_EXECUTABLE=`
which python`"
```
Executing `build.sh` shows that the above expands to 
```
Configuring ...
+ cmake -DMPIEXEC_EXECUTABLE=/apps/slurm/default/bin/srun -DMPIEXEC_NUMPROC_FLAG=-n -DBUILD_GSIBEC=ON -DPYTHON_EXECUTABLE=/scratch1/NCEPDEV/da/python/opt/core/miniconda3/4.6.14/envs/gdasapp/bin/python -DCLONE_JCSDADATA=NO -DWORKFLOW_TESTS=OFF /scratch1/NCEPDEV/da/Russ.Treadon/git/GDASApp/spack-stack/sorc
```
However, later on in the build I see
```
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    PYTHON_EXECUTABLE


-- Build files have been written to: /scratch1/NCEPDEV/da/Russ.Treadon/git/GDASApp/spack-stack/build
```
Looks like I need to add `PYTHON_EXECUTABLE` to `CMakeList.txt` file(s) inside GDASApp.   Is this the right way to go or do we upgrade the version of numpy in the `gdasapp` environment?

```
(gdasapp) Hera(hfe08):/scratch1/NCEPDEV/da/Russ.Treadon/git/GDASApp/spack-stack$ conda list numpy
# packages in environment at /scratch1/NCEPDEV/da/python/opt/core/miniconda3/4.6.14/envs/gdasapp:
#
# Name                    Version                   Build  Channel
numpy                     1.21.6                   pypi_0    pypi
```

CoryMartin-NOAA commented:
Hmm, I'm pretty sure it is because the `py-numpy` module is loaded. Can you remove that and try again?

RussTreadon-NOAA commented:
Unloading `py-numpy/1.22.3` unloads other modules
```
(gdasapp) Hera(hfe08):/scratch1/NCEPDEV/da/Russ.Treadon/git/GDASApp/spack-stack/build$ module unload py-numpy/1.22.3
----------------------------------------------------------------------------------------------------------------------------------
The following dependent module(s) are not currently loaded: py-numpy/1.22.3 (required by: bufr/12.0.1), python/3.10.13 (required by: boost/1.83.0, bufr/12.0.1, fckit/0.11.0, atlas/0.35.1, py-pybind11/2.11.0)
----------------------------------------------------------------------------------------------------------------------------------
```

Rerun `ctest_gdasapp`.  More tests pass.  Now three tests fail
```
90% tests passed, 3 tests failed out of 30

Label Time Summary:
gdas-utils    =   3.34 sec*proc (9 tests)
script        =   3.34 sec*proc (9 tests)

Total Test time (real) =  21.63 sec

The following tests FAILED:
        1692 - test_gdasapp_soca_nsst_increment_to_mom6 (Failed)
        1695 - test_gdasapp_land_apply_jediincr (Failed)
        1696 - test_gdasapp_land_letkfoi_snowda (Failed)
Errors while running CTest
```

RussTreadon-NOAA commented:
Forgot to add the following to my Hera environment
```
export OMP_NUM_THREADS=1
export SLURM_ACCOUNT=da-cpu
export SALLOC_ACCOUNT=$SLURM_ACCOUNT
export SBATCH_ACCOUNT=$SLURM_ACCOUNT
export SLURM_QOS=debug
```
With the above set, only one test fails

```
97% tests passed, 1 tests failed out of 30

Label Time Summary:
gdas-utils    =  17.31 sec*proc (9 tests)
script        =  17.31 sec*proc (9 tests)

Total Test time (real) =  72.02 sec

The following tests FAILED:
        1692 - test_gdasapp_soca_nsst_increment_to_mom6 (Failed)
```
I am still concerned about the other modules unloaded when `py-numpy` is unloaded.

DavidHuber-NOAA commented:
@malloryprow The `gfsmetp*` jobs have finished running on 00z cycles from `2021110800` through `2021111500`.  Could you check the outputs and verify everything is as it should be?  Note that there was a problem archiving `2021110900`, so those `.stat` files are missing, but everything else should be present.  The archives can be found in `/scratch1/NCEPDEV/global/David.Huber/archive/metplus_data/by_VSDB` under `ss_160`.

malloryprow commented:
> @malloryprow The `gfsmetp*` jobs have finished running on 00z cycles from `2021110800` through `2021111500`. Could you check the outputs and verify everything is as it should be? Note that there was a problem archiving `2021110900`, so those `.stat` files are missing, but everything else should be present. The archives can be found in `/scratch1/NCEPDEV/global/David.Huber/archive/metplus_data/by_VSDB` under `ss_160`.

How exciting to see stats being produced in the global-workflow again! The files look fine! Could I see the log files? I want to check on something.

DavidHuber-NOAA commented:
@malloryprow Fantastic!  Sure, the log files are here: `/scratch1/NCEPDEV/global/David.Huber/para/comrot/ss_160/logs`.

CoryMartin-NOAA commented:
> I am still concerned about the other modules unloaded when `py-numpy` is unloaded.

Me too, @RussTreadon-NOAA . I think this is driving me to want to replace the python env with a new one that uses spack-stack python instead of hpc-stack miniconda, but that won't be trivial. The trivial hack fix is to put in the modulefile a prepend to the PYTHONPATH var that looks in our numpy install first.

malloryprow commented:
> @malloryprow Fantastic! Sure, the log files are here: `/scratch1/NCEPDEV/global/David.Huber/para/comrot/ss_160/logs`.

Thanks! I was curious as where the connector script was pointing model_dir_list for EMC_verif-global to. The default value in [run_verif_global_in_global_workflow.sh#L36](https://github.com/NOAA-EMC/EMC_verif-global/blob/gw_updates/ush/run_verif_global_in_global_workflow.sh#L36) kicked in since model_dir is not defined in config.metp.


It would be good to get model_dir defined in config.metp. Would that be okay for this PR? It be adding one line in config.metp on line 24, like in https://github.com/malloryprow/global-workflow/blob/4acc025cdcacdba4ac85c647420a5577854deac3/parm/config/gfs/config.metp.
`export model_dir=${ARCDIR}/..`

DavidHuber-NOAA commented:
@malloryprow Sure, I can put that in.  I'll let it run through another cycle so you can check the logs and `.stat` files again.

RussTreadon-NOAA commented:
@CoryMartin-NOAA and @DavidHuber-NOAA , I added `unload("py-numpy/1.22.3")` to Dave's `hera.lua`.  I also removed the `setenv(R2D2_CONFIG)` line since this is no longer needed.  After this I built the head of GDASApp `develop` and ran the ctests.  All 30 tests pass. 

```
(gdasapp) Hera(hfe10):/scratch1/NCEPDEV/da/Russ.Treadon/git/GDASApp/spack-stack/build$ ctest -R test_gdasapp
Test project /scratch1/NCEPDEV/da/Russ.Treadon/git/GDASApp/spack-stack/build
      Start 1340: test_gdasapp_util_coding_norms
 1/30 Test #1340: test_gdasapp_util_coding_norms .............   Passed    1.25 sec
      Start 1341: test_gdasapp_util_ioda_example
 2/30 Test #1341: test_gdasapp_util_ioda_example .............   Passed    7.30 sec
      Start 1342: test_gdasapp_util_prepdata
 3/30 Test #1342: test_gdasapp_util_prepdata .................   Passed    0.91 sec
      Start 1343: test_gdasapp_util_rads2ioda
 4/30 Test #1343: test_gdasapp_util_rads2ioda ................   Passed    0.40 sec
      Start 1344: test_gdasapp_util_ghrsst2ioda
 5/30 Test #1344: test_gdasapp_util_ghrsst2ioda ..............   Passed    0.14 sec
      Start 1345: test_gdasapp_util_smap2ioda
 6/30 Test #1345: test_gdasapp_util_smap2ioda ................   Passed    0.14 sec
      Start 1346: test_gdasapp_util_smos2ioda
 7/30 Test #1346: test_gdasapp_util_smos2ioda ................   Passed    0.14 sec
      Start 1347: test_gdasapp_util_viirsaod2ioda
 8/30 Test #1347: test_gdasapp_util_viirsaod2ioda ............   Passed    0.14 sec
      Start 1348: test_gdasapp_util_icecamsr2ioda
 9/30 Test #1348: test_gdasapp_util_icecamsr2ioda ............   Passed    0.16 sec
      Start 1680: test_gdasapp_check_python_norms
10/30 Test #1680: test_gdasapp_check_python_norms ............   Passed    1.36 sec
      Start 1681: test_gdasapp_check_yaml_keys
11/30 Test #1681: test_gdasapp_check_yaml_keys ...............   Passed    0.32 sec
      Start 1682: test_gdasapp_jedi_increment_to_fv3
12/30 Test #1682: test_gdasapp_jedi_increment_to_fv3 .........   Passed    1.24 sec
      Start 1683: test_gdasapp_convert_ewok_yaml
13/30 Test #1683: test_gdasapp_convert_ewok_yaml .............   Passed    0.21 sec
      Start 1684: test_gdasapp_convert_bufr_temp_dbuoy
14/30 Test #1684: test_gdasapp_convert_bufr_temp_dbuoy .......   Passed    1.12 sec
      Start 1685: test_gdasapp_convert_bufr_salt_dbuoy
15/30 Test #1685: test_gdasapp_convert_bufr_salt_dbuoy .......   Passed    0.40 sec
      Start 1686: test_gdasapp_convert_bufr_temp_mbuoyb
16/30 Test #1686: test_gdasapp_convert_bufr_temp_mbuoyb ......   Passed    0.18 sec
      Start 1687: test_gdasapp_convert_bufr_salt_mbuoyb
17/30 Test #1687: test_gdasapp_convert_bufr_salt_mbuoyb ......   Passed    0.18 sec
      Start 1688: test_gdasapp_convert_bufr_tesacprof
18/30 Test #1688: test_gdasapp_convert_bufr_tesacprof ........   Passed    0.19 sec
      Start 1689: test_gdasapp_convert_bufr_trkobprof
19/30 Test #1689: test_gdasapp_convert_bufr_trkobprof ........   Passed    0.19 sec
      Start 1690: test_gdasapp_convert_bufr_sfcships
20/30 Test #1690: test_gdasapp_convert_bufr_sfcships .........   Passed    0.19 sec
      Start 1691: test_gdasapp_convert_bufr_sfcshipsu
21/30 Test #1691: test_gdasapp_convert_bufr_sfcshipsu ........   Passed    0.19 sec
      Start 1692: test_gdasapp_soca_nsst_increment_to_mom6
22/30 Test #1692: test_gdasapp_soca_nsst_increment_to_mom6 ...   Passed    8.15 sec
      Start 1693: test_gdasapp_land_create_ens
23/30 Test #1693: test_gdasapp_land_create_ens ...............   Passed    0.83 sec
      Start 1694: test_gdasapp_land_imsproc
24/30 Test #1694: test_gdasapp_land_imsproc ..................   Passed    4.29 sec
      Start 1695: test_gdasapp_land_apply_jediincr
25/30 Test #1695: test_gdasapp_land_apply_jediincr ...........   Passed    4.96 sec
      Start 1696: test_gdasapp_land_letkfoi_snowda
26/30 Test #1696: test_gdasapp_land_letkfoi_snowda ...........   Passed   16.84 sec
      Start 1697: test_gdasapp_convert_bufr_adpsfc_snow
27/30 Test #1697: test_gdasapp_convert_bufr_adpsfc_snow ......   Passed    3.41 sec
      Start 1698: test_gdasapp_convert_bufr_adpsfc
28/30 Test #1698: test_gdasapp_convert_bufr_adpsfc ...........   Passed    4.08 sec
      Start 1699: test_gdasapp_convert_gsi_satbias
29/30 Test #1699: test_gdasapp_convert_gsi_satbias ...........   Passed    1.29 sec
      Start 1700: test_gdasapp_aero_gen_3dvar_yaml
30/30 Test #1700: test_gdasapp_aero_gen_3dvar_yaml ...........   Passed    0.30 sec

100% tests passed, 0 tests failed out of 30

Label Time Summary:
gdas-utils    =  10.57 sec*proc (9 tests)
script        =  10.57 sec*proc (9 tests)

Total Test time (real) =  61.04 sec
```
Not sure if this is the correct solution.

DavidHuber-NOAA commented:
@malloryprow Alright another GFS cycle completed.  Could you check the `.stat` and log files for the `2021111600` cycle?

CoryMartin-NOAA commented:
@RussTreadon-NOAA if it builds, and if the 'workflow tests' pass, i.e. the ones that run the JEDI executables, then I think it might be fine.

malloryprow commented:
> @malloryprow Alright another GFS cycle completed. Could you check the `.stat` and log files for the `2021111600` cycle?

I see model_dir defined through config.metp in all 3 logs (gfsmetpg2g1.log, gfsmetpg2o1.log, gfsmetppcp1.log), and the stat files look good.

Since everything looks good, I would like to merge the branch these changes are in [EMC_verif-global:gw_updates](https://github.com/NOAA-EMC/EMC_verif-global/tree/gw_updates) into [EMC_verif-global:develop](https://github.com/NOAA-EMC/EMC_verif-global/tree/develop). I think that may need to update the branch or commit that the global-workflow is checking out? Is that right @DavidHuber-NOAA?

DavidHuber-NOAA commented:
@malloryprow Yes, that's correct.  I will update it once develop is updated.

RussTreadon-NOAA commented:
> @RussTreadon-NOAA if it builds, and if the 'workflow tests' pass, i.e. the ones that run the JEDI executables, then I think it might be fine.

@DavidHuber-NOAA , my modified copy of your `hera.lua` is in `/scratch1/NCEPDEV/da/Russ.Treadon/git/GDASApp/spack-stack/modulefiles/GDAS`.  Would you commit this to your forked branch, `DavidHuber-NOAA:ss160`.

malloryprow commented:
> @malloryprow Yes, that's correct. I will update it once develop is updated.

It is in [EMC_verif-global:develop](https://github.com/NOAA-EMC/EMC_verif-global/tree/develop) at https://github.com/NOAA-EMC/EMC_verif-global/commit/3dad0a6a72e08f7636b7828ec95df9115ad55157.

DavidHuber-NOAA commented:
@RussTreadon-NOAA Yes, will do, thanks for the leg work!

@malloryprow Great, I'll update the hash.  Thanks!

DavidHuber-NOAA committed:
Assign fcsthrs for awips_g2 job.

DavidHuber-NOAA committed:
Reenable verif-global support

DavidHuber-NOAA committed:
Merge branch 'ss160' of github.com:DavidHuber-NOAA/global-workflow into ss160

RussTreadon-NOAA commented:
Build GDASApp `develop` at [`839e8a2`](https://github.com/NOAA-EMC/GDASApp/commit/839e8a203734b352bed21229ab61a96b591815a9) within working copy g-w branch `DavidHuber-NOAA:ss160` on Hera.  Use modified `gdas.cd/modulefiles/GDAS/hera.lua` discussed [above](https://github.com/NOAA-EMC/global-workflow/pull/2239#issuecomment-1908414764).  Execute g-w GDASApp ctests.  All 52 tests pass on Hera.
```
(gdasapp) Hera(hfe10):/scratch1/NCEPDEV/da/Russ.Treadon/git/global-workflow/spack-stack160/sorc/gdas.cd/build$ ctest -R test_gdasapp       Test project /scratch1/NCEPDEV/da/Russ.Treadon/git/global-workflow/spack-stack160/sorc/gdas.cd/build
      Start 1340: test_gdasapp_util_coding_norms
 1/52 Test #1340: test_gdasapp_util_coding_norms ........................   Passed    1.36 sec
      Start 1341: test_gdasapp_util_ioda_example
 2/52 Test #1341: test_gdasapp_util_ioda_example ........................   Passed   10.41 sec
      Start 1342: test_gdasapp_util_prepdata
 3/52 Test #1342: test_gdasapp_util_prepdata ............................   Passed    1.10 sec
      Start 1343: test_gdasapp_util_rads2ioda
 4/52 Test #1343: test_gdasapp_util_rads2ioda ...........................   Passed    0.41 sec
      Start 1344: test_gdasapp_util_ghrsst2ioda
 5/52 Test #1344: test_gdasapp_util_ghrsst2ioda .........................   Passed    0.18 sec
      Start 1345: test_gdasapp_util_smap2ioda
 6/52 Test #1345: test_gdasapp_util_smap2ioda ...........................   Passed    0.16 sec
      Start 1346: test_gdasapp_util_smos2ioda
 7/52 Test #1346: test_gdasapp_util_smos2ioda ...........................   Passed    0.16 sec
      Start 1347: test_gdasapp_util_viirsaod2ioda
 8/52 Test #1347: test_gdasapp_util_viirsaod2ioda .......................   Passed    0.14 sec
      Start 1348: test_gdasapp_util_icecamsr2ioda
 9/52 Test #1348: test_gdasapp_util_icecamsr2ioda .......................   Passed    0.19 sec
      Start 1680: test_gdasapp_check_python_norms
10/52 Test #1680: test_gdasapp_check_python_norms .......................   Passed    1.46 sec
      Start 1681: test_gdasapp_check_yaml_keys
11/52 Test #1681: test_gdasapp_check_yaml_keys ..........................   Passed    0.16 sec
      Start 1682: test_gdasapp_jedi_increment_to_fv3
12/52 Test #1682: test_gdasapp_jedi_increment_to_fv3 ....................   Passed    1.31 sec
      Start 1683: test_gdasapp_convert_ewok_yaml
13/52 Test #1683: test_gdasapp_convert_ewok_yaml ........................   Passed    0.20 sec
      Start 1684: test_gdasapp_setup_cycled_exp
14/52 Test #1684: test_gdasapp_setup_cycled_exp .........................   Passed    0.54 sec
      Start 1685: test_gdasapp_convert_bufr_temp_dbuoy
15/52 Test #1685: test_gdasapp_convert_bufr_temp_dbuoy ..................   Passed    0.78 sec
      Start 1686: test_gdasapp_convert_bufr_salt_dbuoy
16/52 Test #1686: test_gdasapp_convert_bufr_salt_dbuoy ..................   Passed    0.19 sec
      Start 1687: test_gdasapp_convert_bufr_temp_mbuoyb
17/52 Test #1687: test_gdasapp_convert_bufr_temp_mbuoyb .................   Passed    0.23 sec
      Start 1688: test_gdasapp_convert_bufr_salt_mbuoyb
18/52 Test #1688: test_gdasapp_convert_bufr_salt_mbuoyb .................   Passed    0.18 sec
      Start 1689: test_gdasapp_convert_bufr_tesacprof
19/52 Test #1689: test_gdasapp_convert_bufr_tesacprof ...................   Passed    0.19 sec
      Start 1690: test_gdasapp_convert_bufr_trkobprof
20/52 Test #1690: test_gdasapp_convert_bufr_trkobprof ...................   Passed    0.18 sec
      Start 1691: test_gdasapp_convert_bufr_sfcships
21/52 Test #1691: test_gdasapp_convert_bufr_sfcships ....................   Passed    0.19 sec
      Start 1692: test_gdasapp_convert_bufr_sfcshipsu
22/52 Test #1692: test_gdasapp_convert_bufr_sfcshipsu ...................   Passed    0.20 sec
      Start 1693: test_gdasapp_soca_nsst_increment_to_mom6
23/52 Test #1693: test_gdasapp_soca_nsst_increment_to_mom6 ..............   Passed   10.06 sec
      Start 1694: test_gdasapp_soca_prep
24/52 Test #1694: test_gdasapp_soca_prep ................................   Passed    2.93 sec
      Start 1695: test_gdasapp_soca_run_clean
25/52 Test #1695: test_gdasapp_soca_run_clean ...........................   Passed    0.07 sec
      Start 1696: test_gdasapp_soca_setup_obsprep
26/52 Test #1696: test_gdasapp_soca_setup_obsprep .......................   Passed    0.49 sec
      Start 1697: test_gdasapp_soca_JGLOBAL_PREP_OCEAN_OBS
27/52 Test #1697: test_gdasapp_soca_JGLOBAL_PREP_OCEAN_OBS ..............   Passed   10.87 sec
      Start 1698: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP
28/52 Test #1698: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP ....   Passed   43.06 sec
      Start 1699: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_BMAT
29/52 Test #1699: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_BMAT ....   Passed   74.23 sec
      Start 1700: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_RUN
30/52 Test #1700: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_RUN .....   Passed   42.23 sec
      Start 1701: test_gdasapp_soca_copy_scratch
31/52 Test #1701: test_gdasapp_soca_copy_scratch ........................   Passed    0.73 sec
      Start 1702: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_CHKPT
32/52 Test #1702: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_CHKPT ...   Passed   42.83 sec
      Start 1703: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_POST
33/52 Test #1703: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_POST ....   Passed   10.51 sec
      Start 1704: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_VRFY
34/52 Test #1704: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_VRFY ....   Passed  299.61 sec
      Start 1705: test_gdasapp_soca_socahybridweights
35/52 Test #1705: test_gdasapp_soca_socahybridweights ...................   Passed   42.31 sec
      Start 1706: test_gdasapp_soca_incr_handler
36/52 Test #1706: test_gdasapp_soca_incr_handler ........................   Passed   10.19 sec
      Start 1707: test_gdasapp_soca_ens_handler
37/52 Test #1707: test_gdasapp_soca_ens_handler .........................   Passed   10.19 sec
      Start 1708: test_gdasapp_land_create_ens
38/52 Test #1708: test_gdasapp_land_create_ens ..........................   Passed    1.33 sec
      Start 1709: test_gdasapp_land_imsproc
39/52 Test #1709: test_gdasapp_land_imsproc .............................   Passed    4.92 sec
      Start 1710: test_gdasapp_land_apply_jediincr
40/52 Test #1710: test_gdasapp_land_apply_jediincr ......................   Passed    4.38 sec
      Start 1711: test_gdasapp_land_letkfoi_snowda
41/52 Test #1711: test_gdasapp_land_letkfoi_snowda ......................   Passed   10.84 sec
      Start 1712: test_gdasapp_convert_bufr_adpsfc_snow
42/52 Test #1712: test_gdasapp_convert_bufr_adpsfc_snow .................   Passed    4.23 sec
      Start 1713: test_gdasapp_convert_bufr_adpsfc
43/52 Test #1713: test_gdasapp_convert_bufr_adpsfc ......................   Passed    4.04 sec
      Start 1714: test_gdasapp_convert_gsi_satbias
44/52 Test #1714: test_gdasapp_convert_gsi_satbias ......................   Passed    1.28 sec
      Start 1715: test_gdasapp_setup_atm_cycled_exp
45/52 Test #1715: test_gdasapp_setup_atm_cycled_exp .....................   Passed    0.71 sec
      Start 1716: test_gdasapp_atm_jjob_var_init
46/52 Test #1716: test_gdasapp_atm_jjob_var_init ........................   Passed   43.73 sec
      Start 1717: test_gdasapp_atm_jjob_var_run
47/52 Test #1717: test_gdasapp_atm_jjob_var_run .........................   Passed  106.12 sec
      Start 1718: test_gdasapp_atm_jjob_var_final
48/52 Test #1718: test_gdasapp_atm_jjob_var_final .......................   Passed   42.12 sec
      Start 1719: test_gdasapp_atm_jjob_ens_init
49/52 Test #1719: test_gdasapp_atm_jjob_ens_init ........................   Passed   43.68 sec
      Start 1720: test_gdasapp_atm_jjob_ens_run
50/52 Test #1720: test_gdasapp_atm_jjob_ens_run .........................   Passed  266.15 sec
      Start 1721: test_gdasapp_atm_jjob_ens_final
51/52 Test #1721: test_gdasapp_atm_jjob_ens_final .......................   Passed   74.54 sec
      Start 1722: test_gdasapp_aero_gen_3dvar_yaml
52/52 Test #1722: test_gdasapp_aero_gen_3dvar_yaml ......................   Passed    0.42 sec

100% tests passed, 0 tests failed out of 52

Label Time Summary:
gdas-utils    =  14.10 sec*proc (9 tests)
script        =  14.10 sec*proc (9 tests)

Total Test time (real) = 1230.75 sec
```

DavidHuber-NOAA committed:
Update gdasapp to include ss/1.6.0 support.

DavidHuber-NOAA committed:
New (cleaner history) gdas hash.

DavidHuber-NOAA commented:
@RussTreadon-NOAA @CoryMartin-NOAA Thanks for the tests!  Should the Hercules and Orion modules be updated in the same fashion?  I can handle that and do test builds on both systems if so.

CoryMartin-NOAA commented:
I believe so yes, thanks!

DavidHuber-NOAA committed:
Update hashes (revert WCOSS2 modules).

RussTreadon-NOAA commented:
@DavidHuber-NOAA , do we want to create a GDASApp spack-stack 1.6.0 [PR](https://github.com/NOAA-EMC/GDASApp/pulls) and add it to the list of dependencies for this PR?

DavidHuber-NOAA commented:
@RussTreadon-NOAA Yes, that's the plan.  I successfully tested a build on Hercules yesterday and will work on Orion today.  Once complete, I'll open the PR.

RussTreadon-NOAA commented:
@DavidHuber-NOAA , excellent!  Thank you very much for including GDASApp in this PR.

DavidHuber-NOAA committed:
Fix wgrib2/gfs_utils on Hercules.

DavidHuber-NOAA committed:
Merge branch 'ss160' of github.com:DavidHuber-NOAA/global-workflow into ss160

DavidHuber-NOAA commented:
This PR is blocked pending the replacement of the wrong `amsua_metop-c.SpcCoeff.bin` CRTM-fix coefficient file (https://github.com/JCSDA/spack-stack/issues/963).

Also updated the dependency list in the description with additional PRs for the UPP and UFS_Utils.

DavidHuber-NOAA committed:
Update ufs_utils hash for spack-stack/1.6.0 support.

DavidHuber-NOAA committed:
Merge branch 'ss160' of github.com:DavidHuber-NOAA/global-workflow into ss160

DavidHuber-NOAA committed:
Update GDAS hash to include SS/1.6.0 support.

DavidHuber-NOAA committed:
Update Orion, Hercules, S4 modulefiles.

DavidHuber-NOAA committed:
Update monitor hash to noaa-emc with SS/1.6.0 support.

DavidHuber-NOAA committed:
Fix gfs_utils Orion spack-stack env path.

DavidHuber-NOAA committed:
Merge branch 'ss160' of github.com:DavidHuber-NOAA/global-workflow into ss160

DavidHuber-NOAA commented:
After incorporating the most recent UFS_Utils hash to include spack-stack/1.6.0 support, `global_cycle` (in `gdassfcanl` and `gfssfcanl`) are failing as the expected namelist inputs have changed.  I believe this will be resolved by #2262.

CatherineThomas-NOAA commented:
@DavidHuber-NOAA Can you point me to a log file for the failed global_cycle?  
cc @GeorgeGayno-NOAA

DavidHuber-NOAA commented:
@CatherineThomas-NOAA Yes, you can find it here: `/work2/noaa/global/dhuber/para/com/herc_160_new_f2o/logs/2021110800/gdassfcanl.log`.

CatherineThomas-NOAA commented:
Thanks @DavidHuber-NOAA.  As you said, it's complaining about the frac_grid variable in the namelist, but if the UFS_UTILS hash you are pointing to is [547459c](https://github.com/ufs-community/UFS_UTILS/commit/547459c1e323ca45854d7ac1e1cb11e73f8de408), the fractional grid changes should already be included and should not need PR #2262 to work.

Looking at /work/noaa/global/dhuber/GW/ss_160_herc/sorc/ufs_utils.fd/sorc/global_cycle.fd/cycle.f90, I see FRAC_GRID listed:
```
 NAMELIST/NAMCYC/ IDIM,JDIM,LSOIL,LUGB,IY,IM,ID,IH,FH,&
                  DELTSFC,IALB,USE_UFO,DONST,             &
                  DO_SFCCYCLE,ISOT,IVEGSRC,ZSEA1_MM,      &
                  ZSEA2_MM, MAX_TASKS, DO_LNDINC, FRAC_GRID
```

I'm not sure why this is failing.  @GeorgeGayno-NOAA Can you double check me?

DavidHuber-NOAA commented:
@CatherineThomas-NOAA Thanks for checking.  I suspect that I did not recompile global_cycle.  Let me give that a try and see if that fixes the issue.

DavidHuber-NOAA commented:
@CatherineThomas-NOAA Sure enough, recompiling resolved the issue.  Thanks again.

CatherineThomas-NOAA commented:
@DavidHuber-NOAA Glad it was an easy fix!

DavidHuber-NOAA commented:
The CRTM-fix files have been updated on all supported platforms and have been tested on Orion, Jet, Hera, and Hercules, so this PR is no longer blocked by that.

Cycle testing has been completed on Orion, Hercules, and Hera.

There are two major hurdle left.  One is updating the UFS to spack-stack/1.6.0 and the other is the GSI RTs.  The GSI RTs pass on all platforms except Hercules where netCDF/HDF errors are being generated for 4 tests.  See PR https://github.com/NOAA-EMC/GSI/pull/684 for more details.

henrywinterbottom-wxdev committed:
Removed jlogfile and postmsg references within gempak scripts.

henrywinterbottom-wxdev committed:
Removed dummy variable passed to perl scripts.

henrywinterbottom-wxdev committed:
Removed jlogfile references.

HenryRWinterbottom committed:
Merge branch 'NOAA-EMC:develop' into feature/gfsv17_issue_2125

DavidHuber-NOAA committed:
Revert Hercules modules to SS/1.5.1.

DavidHuber-NOAA committed:
Fix SS versions for CI modules.

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into ss160

HenryRWinterbottom committed:
Merge branch 'NOAA-EMC:develop' into feature/gfsv17_issue_2125

henrywinterbottom-wxdev committed:
Implemented ush/detect_machine.sh for host determination and removed redundant checks for expected file paths.

HenryRWinterbottom committed:
Merge branch 'NOAA-EMC:develop' into feature/gfsv17_issue_2125

HenryRWinterbottom committed:
Merge branch 'NOAA-EMC:develop' into feature/gwdev_issue_2129

henrywinterbottom-wxdev committed:
Corrected based on reviewer review.

henrywinterbottom-wxdev committed:
Updates based on user request.

DavidHuber-NOAA committed:
Update GSI hashes.

DavidHuber-NOAA committed:
Reenable METplus jobs on Hercules.

DavidHuber-NOAA committed:
Reupgrade Hercules to SS/1.6.0

DavidHuber-NOAA committed:
Merge remote-tracking branch 'origin/develop' into ss160

DavidHuber-NOAA committed:
Updated GSI-Utils hash to head of develop.

DavidHuber-NOAA commented:
@aerorahul @WalterKolczynski-NOAA I would like to go ahead and move forward with this PR even though the UFS isn't quite ready yet.  Since UFS modules are loaded separately anyway via `load_fv3_modules.sh`, I think this should be fine.  This will allow the CRTM, DA, and METplus updates to be incorporated into the global workflow.  We can then update the UFS hash when it is ready in a separate PR.  Any objections to that idea?  If not, then I will start final testing on Hera, Orion, and Hercules after repointing the UFS hash to `a837cfc` (our current hash in develop).

DavidHuber-NOAA committed:
Fixed xarray version for SS/1.6.0.

DavidHuber-NOAA committed:
Fix git version in Hera's gwsetup module.

DavidHuber-NOAA committed:
Revert UFS hash.

DavidHuber-NOAA committed:
Fixed Orion cdo version.

HenryRWinterbottom committed:
Merge branch 'NOAA-EMC:develop' into feature/gwdev_issue_2129

DavidHuber-NOAA committed:
Load default rocoto on Jet.

HenryRWinterbottom committed:
Merge branch 'NOAA-EMC:develop' into feature/gwdev_issue_2129

DavidHuber-NOAA commented:
@KateFriedman-NOAA @malloryprow I am looking at enabling the METplus jobs on WCOSS2, but this will require appending to the `MODULEPATH` in module_base.wcoss2.lua in order to load the `met/9.1.3` and `metplus/3.1.1` modules.  I seem to recall appending to the `MODULEPATH` is a no-no on WCOSS2.  Is that correct?

malloryprow commented:
@DavidHuber-NOAA I remember things being weird with trying to load MET and METplus on WCOSS2 because the versions needed are in para. Is that what you are talking about having to add to MODULEPATH?

KateFriedman-NOAA commented:
> I seem to recall appending to the `MODULEPATH` is a no-no on WCOSS2. Is that correct?

@DavidHuber-NOAA I don't recall that being a no-no on WCOSS2 but we can check on that. We're already appending in our current `module_base.wcoss2.lua`:

https://github.com/NOAA-EMC/global-workflow/blob/develop/modulefiles/module_base.wcoss2.lua#L41

If that is a no-no for ops then we shouldn't need to worry because we'll likely build an ops-specific modulefile (or have massively reconfigured how we load modules by then) and we won't be loading dev-only things like `met` or `metplus` in the ops one.

I say add it as an append for now and we can sort out potential issues are we prep for the ops configuration. Thanks!

DavidHuber-NOAA commented:
@malloryprow Yes, that correct.

@KateFriedman-NOAA Okie dokie, thanks!

HenryRWinterbottom committed:
Merge branch 'NOAA-EMC:develop' into feature/gwdev_issue_2129

HenryRWinterbottom committed:
Merge branch 'NOAA-EMC:develop' into feature/gfsv17_issue_2125

DavidHuber-NOAA commented:
Hmm, I'm having some trouble with WCOSS2.  After updating `run_verif_global_in_global_workflow.sh` to add the `HOMEMETplus` variable, it looks like it is having trouble with the `gfsmetpg2o1` and `gfsmetppcp1` jobs.  I'll create a verif-global issue to track this and circle back around to it in #2106.

malloryprow commented:
I had it working on WCOSS2 before. I had to add some changes (https://github.com/NOAA-EMC/EMC_verif-global/blob/babaa3638f9b181c399284fe443b90418baf76f2/ush/run_verif_global_in_global_workflow.sh#L230). Maybe the workaround I added is now causing the problems.

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into ss160

DavidHuber-NOAA committed:
Merge branch 'ss160' of github.com:DavidHuber-NOAA/global-workflow into ss160

henrywinterbottom-wxdev committed:
Linter corrections.

henrywinterbottom-wxdev committed:
Updates requested by reviewer; testing -- DO NOT REVIEW.

henrywinterbottom-wxdev committed:
Updates requested by reviewer Rahul Mahajan.

henrywinterbottom-wxdev committed:
Bug fix.

Commit:
Added verif-global support back to WCOSS2.

DavidHuber-NOAA commented:
This branch has been tested on Hercules, Orion, Hera, and WCOSS2.  It works on all of these systems and I am thus marking this ready for review.

aerorahul commented on parm/config/gfs/config.base.emc.dyn:
```diff
@@ -377,7 +377,7 @@ export netcdf_diag=".true."
 
 # Verification options
-export DO_METP="NO"          # Run METPLUS jobs - set METPLUS settings in config.metp; not supported with spack-stack
+export DO_METP="YES"         # Run METPLUS jobs - set METPLUS settings in config.metp; not supported with spack-stack
```
I think the comment about not supported w/ spack-stack can be removed now.

aerorahul commented on sorc/build_upp.sh:
```diff
@@ -6,10 +6,11 @@ cd "${script_dir}" || exit 1
 	case "${option}" in
 		d) _opts+="-d ";;
 		v) _opts+="-v ";;
+      j) export BUILD_JOBS="${OPTARG}"; echo success ;;
```
```suggestion
    j) export BUILD_JOBS="${OPTARG}"; echo success ;;
```

aerorahul commented on versions/build.hera.ver:
```diff
@@ -1,3 +1,4 @@
 export stack_intel_ver=2021.5.0
 export stack_impi_ver=2021.5.1
+export spack_env=gsi-addon-dev
```
is it possible to put a variable here such as:
```bash
export spack_mod_path="/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev/install/modulefiles/Core"
```
and then in the `module_base.hera.lua` have something like:
```lua
local spack_mod_path=os.getenv('spack_mod_path') or "None"
prepend_path("MODULEPATH", spack_mod_path)
```

DavidHuber-NOAA replied:
Yes, I think so -- it sounds a lot less ugly, too.  I'll give that a try.

DavidHuber-NOAA replied:
Fixed and tested on all research systems.

aerorahul reviewed (commented):
changes look good.
just a comment and an observation.

DavidHuber-NOAA committed:
Fix whitespace in build_upp.sh.

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

WalterKolczynski-NOAA commented on sorc/build_upp.sh:
```diff
@@ -6,10 +6,11 @@ cd "${script_dir}" || exit 1
 	case "${option}" in
 		d) _opts+="-d ";;
 		v) _opts+="-v ";;
+    j) export BUILD_JOBS="${OPTARG}"; echo success ;;
```
Indentation does not match and I'd prefer these in alphabetical order.

DavidHuber-NOAA replied:
Same mistake twice... I think I have it now.

DavidHuber-NOAA committed:
Clean up build_upp.sh.

-Corrected whitespace (tabs instead of spaces)
-Removed debug print statement
-Alphebetized flags

DavidHuber-NOAA committed:
Update verif-global to latest WCOSS2 support.

KateFriedman-NOAA commented on modulefiles/module_base.hera.lua:
```diff
@@ -33,6 +33,10 @@ load(pathJoin("wgrib2", (os.getenv("wgrib2_ver") or "None")))
+load(pathJoin("metplus", (os.getenv("metplus_ver") or "None")))
 load(pathJoin("py-xarray", (os.getenv("py_xarray_ver") or "None")))
 
 -- MET/METplus are not available for use with spack-stack, yet
```
Can remove these commented out MET/METPLUS lines now.

KateFriedman-NOAA commented on modulefiles/module_base.orion.lua:
```diff
@@ -31,6 +31,10 @@ load(pathJoin("wgrib2", (os.getenv("wgrib2_ver") or "None")))
+load(pathJoin("metplus", (os.getenv("metplus_ver") or "None")))
 load(pathJoin("py-xarray", (os.getenv("py_xarray_ver") or "None")))
 
 -- MET/METplus are not yet supported with spack-stack
```
Can remove these commented out MET/METPLUS lines now.

DavidHuber-NOAA committed:
Update comment about METplus support.

DavidHuber-NOAA committed:
Move SS module path to version files.

KateFriedman-NOAA commented on versions/build.hercules.ver:
```diff
@@ -1,3 +1,5 @@
+export spack_mod_path="/work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.6.0/envs/gsi-addon-env/install/modulefiles/Core"
```
Could these `spack_mod_path` variables use `spack_stack_ver` and `spack_env` variables so we can reduce the places we'd have to update those when they change?

KateFriedman-NOAA replied:
Thanks for making the change! Looks good!

KateFriedman-NOAA reviewed (changes requested):
Looks good, thanks for this big update @DavidHuber-NOAA ! Left a few comments to consider before I approve.

DavidHuber-NOAA committed:
Removed MET/METplus 'not available' comments

DavidHuber-NOAA committed:
Construct SS paths from version variables.

KateFriedman-NOAA reviewed (approved):
Looks great, thanks @DavidHuber-NOAA !

DavidHuber-NOAA commented:
@souopgui reported that this PR ran successfully on S4 through 9 cycles, including the generation of verif-global .stat files.

DavidHuber-NOAA committed:
Update GDASApp hash to include SS/1.6.0 support.

DavidHuber-NOAA commented:
I updated the PR description to include the auxiliary/coincidental work included in this PR.

KateFriedman-NOAA commented:
> I updated the PR description to include the auxiliary/coincidental work included in this PR.

Thanks @DavidHuber-NOAA ! Updated description looks great!

aerorahul reviewed (approved):
LGTM.
Thanks for the updates.

WalterKolczynski-NOAA reviewed (approved):
Conditionally approved pending successful completion of CI tests

aerorahul commented:
> Automated global-workflow Testing Results:
> 
> ```
> Machine: Hera
> Start: Wed Feb 21 18:25:22 UTC 2024 on hfe05
> ---------------------------------------------------
> Build: *** FAILED ***
> Build: Failed at 02/21/24 06:43:52 PM
> Creating logs folder
> Creating ../exec folder
> Resetting modules to system default. Reseting $MODULEPATH back to system default. All extra directories will be removed from $MODULEPATH.
> Building gsi_enkf, gdas, ww3prepost, ufs_utils, ufs, gsi_utils, gsi_monitor, gfs_utils, upp
> Starting build_gsi_enkf.sh
> Starting build_gdas.sh
> Starting build_ww3prepost.sh
> Starting build_ufs_utils.sh
> Starting build_ufs.sh
> Starting build_gsi_utils.sh
> Starting build_gsi_monitor.sh
> Starting build_gfs_utils.sh
> Starting build_upp.sh
> build_gdas.sh failed with status 1!
> build_ufs_utils.sh completed successfully!
> build_gsi_utils.sh completed successfully!
> build_gfs_utils.sh completed successfully!
> build_gsi_monitor.sh completed successfully!
> build_ww3prepost.sh completed successfully!
> build_upp.sh completed successfully!
> build_gsi_enkf.sh completed successfully!
> build_ufs.sh completed successfully!
> BUILD ERROR: One or more components failed to build
>   Check the associated build log(s) for details.
> Failed on cloning and building global-workflowi PR: 2239
> CI on Hera failed to build on Wed Feb 21 18:43:53 UTC 2024 for repo https://github.com/NOAA-EMC/global-workflow.git
> ```

@DavidHuber-NOAA 
The Hera build is failing to build GDASApp on Hera.
`/scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/2239/global-workflow`
and `sorc/logs/build_gdas.log`

It is trying to load rocoto/1.3.3 which is not available.  It is possible that the GDASApp hash in this PR is stale and updating it will fix the issue and load rocoto/1.3.6

DavidHuber-NOAA committed:
Updated GDAS to include rocoto/1.3.6 on Hera.

DavidHuber-NOAA commented:
@aerorahul Rocoto was updated to 1.3.6 just after we (with Cory and Russ) and was updated 5 days later.  I updated the hash from https://github.com/NOAA-EMC/GDASApp/commit/9ba6bb9916abf93de899d88a540904572bdbd7a7 to https://github.com/NOAA-EMC/GDASApp/commit/0ebe826a41cb7bf0ea277791060f5bc03892cd3e and ran a test build, which was successful.

The Orion CI failure seems to suggest that `lmod` was not initialized at login.  From the log:

```console
Begin stage_ic.sh at Wed Feb 21 19:25:18 UTC 2024
+ stage_ic.sh[6]: . /work2/noaa/stmp/GFS_CI_ROOT/ORION/PR/2239/global-workflow/ush/load_fv3gfs_modules.sh
++ load_fv3gfs_modules.sh[4]: [[ NO == \N\O ]]
++ load_fv3gfs_modules.sh[5]: echo 'Loading modules quietly...'
Loading modules quietly...
++ load_fv3gfs_modules.sh[6]: set +x
Resetting modules to system default. Reseting $MODULEPATH back to system default. All extra directories will be removed from $MODULEPATH.
Lmod Warning: MODULEPATH is undefined.



Lmod has detected the following error: The following module(s) are unknown:
"contrib"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore-cache load "contrib"

Also make sure that all modulefiles written in TCL start with the string
#%Module



End stage_ic.sh at 19:25:19 with error code 1 (time elapsed: 00:00:01)
```

I will restart the CI tests now.

JessicaMeixner-NOAA commented:
@DavidHuber-NOAA I had the same orion failures on the last PR.  @aerorahul can tell you more but it seems as if the underlying issues are not yet resolved, but it's not your PR causing those failures.

DavidHuber-NOAA commented:
@JessicaMeixner-NOAA Noted, thanks!

DavidHuber-NOAA commented:
@HenryWinterbottom-NOAA @aerorahul I am going to merge #2324 and #2312 into this PR then kick off the CI tests on Hera again.

TerrenceMcGuinness-NOAA commented:
I have an idea about what might be causing the error in CI cron BASH on Orion that may have been introduced from recent updates to the **develop** branch.  Unfortunately,  my Orion account has expired so I can not check myself.  We're looking into getting my account refreshed.

DavidHuber-NOAA committed:
Merge remote-tracking branch 'henry/feature/gfsv17_issue_2125' into ss160

DavidHuber-NOAA committed:
Merge remote-tracking branch 'henry/feature/gwdev_issue_2129' into ss160

DavidHuber-NOAA committed:
Merge remote-tracking branch 'origin/develop' into ss160

TerrenceMcGuinness-NOAA commented:
Rahul has asked me to cancel this CI run (Bash CI can be re-start but not outright canceled - feature pending).  I can't cancel the Orion case due to accounting issues, but I'm canceling the Hera instance.

HenryRWinterbottom commented:
> @HenryWinterbottom-NOAA @aerorahul I am going to merge #2324 and #2312 into this PR then kick off the CI tests on Hera again.

Thank you! Please do.

aerorahul commented:
@DavidHuber-NOAA @WalterKolczynski-NOAA 
Just a note:
When the CI finishes successfully, let us create a tag for HR3 before merging this PR as the testing for HR3 was done w/ SS 1.5.1.  One can create a tag from a previous hash in `develop`, but this way is cleaner and clear.
Wait for a green light on creating a tag from @JessicaMeixner-NOAA as some final checks are being performed.
Thanks!

aerorahul commented:
> ```
> Experiment C96_atmsnowDA_79144f24  *** FAILED *** on Hera
> Experiment C96_atmsnowDA_79144f24  with 1 tasks failed at 02/22/24 05:40:23 PM
> Error logs:
> /scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/2239/RUNTESTS/COMROOT/C96_atmsnowDA_79144f24/logs/2021122018/gdasfcst.log
> ```

This looks like one of those `global-cycle` related updates.
The error message essentially is:
```
 3:
 3: FATAL from PE     3: The checksum in the file:INPUT/sfc_data.tile1.nc and variable:sheleg does not match the checksum calculated from the data. fi
le:DE5D5DEB4C7622A7 from data:D55F574CD77BCA7E
 3:
 0:
```

The `sfc_data.tile1.nc` file going into `global-cycle` for ex. is at:
```
/scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/2239/RUNTESTS/COMROOT/C96_atmsnowDA_79144f24/gdas.20211220/18//analysis/land/20211220.180000.sfc_data.tile1.nc 
```
The output `sfcanl_data.tile1.nc` coming out of `global-cycle` for ex. is at:
```
/scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/2239/RUNTESTS/COMROOT/C96_atmsnowDA_79144f24/gdas.20211220/18//model_data/atmos/restart/20211220.180000.sfcanl_data.tile1.nc
```

@GeorgeGayno-NOAA @jiaruidong2017 Is something obvious to either of you?  Thanks!

jiaruidong2017 commented:
It seems not to include the recent updates of GDASApp (https://github.com/NOAA-EMC/GDASApp/pull/899) by removing the checksum attribute to the modified variables during JEDI based snow DA analysis.

jiaruidong2017 commented:
commit 8b4bca9e8678c8bef41df9c5047f7494cd9daeac
Author: Jiarui Dong <Jiarui.Dong@noaa.gov>
**Date:   Fri Feb 2 10:22:49 2024 -0500**

    Update the commit hash of land-jediincr for update global_cycle in Global-Workflow (#899)

    This PR updates the commit hash from NOAA-PSL/land-apply_jedi_incr to
    include the recent updates
    https://github.com/NOAA-PSL/land-apply_jedi_incr/pull/4, which remove
    the checksum attribute to the modified variables during JEDI based snow
    DA analysis.

    This PR will contribute to
    https://github.com/NOAA-EMC/global-workflow/pull/2262 to get the
    global_cycle into the global-workflow.

    Co-authored-by: Cory Martin <cory.r.martin@noaa.gov>

DavidHuber-NOAA commented:
@jiaruidong2017 @aerorahul Thanks, I'll try updating the GDASApp hash to https://github.com/NOAA-EMC/GDASApp/commit/8b4bca9e8678c8bef41df9c5047f7494cd9daeac.

DavidHuber-NOAA committed:
Update GDAS hash to allow modified snow DA analysis

RussTreadon-NOAA commented:
@jiaruidong2017 , is there any reason we should not update the GDASApp hash to the current head at [`984ecb`](https://github.com/NOAA-EMC/GDASApp/commit/984ecbea3cd971cfb98b4ce7d9299442a38e8dd1)?  This hash was created yesterday, 2/21/2024.  

GDASApp hash [`8b4bca`](https://github.com/NOAA-EMC/GDASApp/commit/8b4bca9e8678c8bef41df9c5047f7494cd9daeac) dates back to 2/2/2024.

jiaruidong2017 commented:
> @jiaruidong2017 , is there any reason we should not update the GDASApp hash to the current head at [`984ecb`](https://github.com/NOAA-EMC/GDASApp/commit/984ecbea3cd971cfb98b4ce7d9299442a38e8dd1)? This hash was created yesterday, 2/21/2024.
> 
> GDASApp hash [`8b4bca`](https://github.com/NOAA-EMC/GDASApp/commit/8b4bca9e8678c8bef41df9c5047f7494cd9daeac) dates back to 2/2/2024.

@RussTreadon-NOAA We definitely can update  the GDASApp hash to the current version.

DavidHuber-NOAA commented:
@jiaruidong2017 @RussTreadon-NOAA Alright, thanks.  I'll bump the hash to current develop.

@TerrenceMcGuinness-NOAA Would you mind killing the Hera CI jobs?
A: "sure thing, CI can do that for us ;)"

DavidHuber-NOAA committed:
Update GDASApp hash to current develop

WalterKolczynski-NOAA commented:
```2024-02-22 20:27:32,653 - INFO     - land_analysis: Reading /scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/2239/global-workflow/sorc/gdas.cd/parm/land/prep/prep_gts.yaml
Traceback (most recent call last):
  File "/scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/2239/global-workflow/scripts/exglobal_prep_land_obs.py", line 23, in <module>
    LandAnl.prepare_GTS()
  File "/scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/2239/global-workflow/ush/python/wxflow/logger.py", line 266, in wrapper
    retval = func(*args, **kwargs)
  File "/scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/2239/global-workflow/ush/python/pygfs/task/land_analysis.py", line 83, in prepare_GTS
    prep_gts_config = parse_j2yaml(self.task_config.GTS_OBS_LIST, localconf)
  File "/scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/2239/global-workflow/ush/python/wxflow/yaml_file.py", line 176, in parse_j2yaml
    yaml_dict = YAMLFile(data=yaml_file)
  File "/scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/2239/global-workflow/ush/python/wxflow/yaml_file.py", line 35, in __init__
    self.update(config)
  File "/scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/2239/global-workflow/ush/python/wxflow/attrdict.py", line 120, in update
    other.update(args[0])
ValueError: dictionary update sequence element #0 has length 1; 2 is required
```

CoryMartin-NOAA commented:
99.9% sure this commit would solve this issue: https://github.com/NOAA-EMC/global-workflow/pull/2330/commits/8cbc974a981a1768d838457640e27f43f582c10b

jiaruidong2017 commented:
Yes, I agree with Cory. This failure was caused by recent merged commit hash by changing from `land` to `snow`.

DavidHuber-NOAA committed:
Disable snow DA tests

DavidHuber-NOAA commented:
Per suggestion from @aerorahul I'm going to proceed without running the snow DA tests.  We can reenable them when the issues have been sorted out in a future PR.

jiaruidong2017 commented:
This is caused by the following commit hash. Cory's new PR #2330 will solve this issue. 

```
commit 0376c0812952dcdfb5ebaf3b04dbd625225a2def
Author: Cory Martin <cory.r.martin@noaa.gov>
Date:   Tue Feb 20 08:41:41 2024 -0500

    Changing 'land' to 'snow' to better reflect analysis jobs (#913)

    This PR changes 'land' everywhere to 'snow' to differentiate that this
    is snow DA and not land DA. There will need to be a corresponding G-W PR
    to do this in the workflow.
```

For now, you can use the early commit hashes as below for your test.

```
commit 49344c9084b26a6289308fc2e4433758542606b5
Author: Guillaume Vernieres <guillaume.vernieres@noaa.gov>
Date:   Sat Feb 17 09:51:18 2024 -0500

    Update to the bkg name convention (#924)

    - fixes #918

```
```
commit f09fec90b56c99ab78a1932975e15996956e6bbc
Author: Guillaume Vernieres <guillaume.vernieres@noaa.gov>
Date:   Fri Feb 16 12:59:59 2024 -0500

    Allow for gdas or gfs obs (#925)

    - fixes #921

```
